### PR TITLE
Disallow modifier declarations and definitions in interfaces

### LINKED ISSUE #11557: Modifiers are not disallowed in interfaces
## Description

Currently it's possible to define modifiers in interfaces and even provide an implementation. This is not documented. [The current section about interfaces](https://docs.soliditylang.org/en/v0.8.5/contracts.html#interfaces) only mentions functions so technically it does not contradict the docs but I think this is not intentional. It does have some quirks, e.g. declaring a local variable causes an error.

Implemented modifiers should be disallowed in interfaces. Unimplemented ones probably too given that modfiers cannot be external.

Regardless of whether we change this or not, docs should be updated to say whether modifiers are allowed in interfaces. I think that we should also mention that at least up to 0.8.5 they were allowed .

## Environment

- Compiler version: 0.8.5

### Implemented modifier in an interface
This compiles without errors:
```solidity
function free() {}

interface I {
    modifier m() {
        free();
        _;
    }
}

contract C is I {
    function f() public m {}
}
```

### Implemented modifier with a local variable in an interface
```solidity
function free() {}

interface I {
    modifier m() {
        uint answer = 42;
        free();
        _;
    }
}

contract C is I {
    function f() public m {}
}
```
```
Error: Variables cannot be declared in interfaces.
 --> test.sol:5:9:
  |
5 |         uint answer = 42;
  |   
```

### Implemented modifier with a local variable in an abstract contract
Local variables work just fine in modifiers in abstract contracts:
```solidity
function free() {}

abstract contract I {
    modifier m() {
        uint answer = 42;
        free();
        _;
    }
}

contract C is I {
    function f() public m {}
}
```

### PR DESCRIPTION
**Pull request by cameel**
Fixes #11557.

### CODE REVIEW

**Comment by hrkrshnn** [comment]
Shouldn't this be breaking?

**Comment by cameel** [comment]
We discussed it on the chat and @chriseth's opinion was that it should not.

**Comment by chriseth** [comment]
Strictly it is breaking. @axic what do you think?

File: test/libsolidity/analysis/FunctionCallGraph.cpp [hunk 1]
```diff
@@ -1114,7 +1114,6 @@ BOOST_AUTO_TEST_CASE(interfaces_and_abstract_contracts)
 	unique_ptr<CompilerStack> compilerStack = parseAndAnalyzeContracts(R"(
 		interface I {
 			event Ev(uint);
-			modifier m() virtual;
```
Comment (axic) [inline test/libsolidity/analysis/FunctionCallGraph.cpp:1117]: Damn, we even had a test for it?!

**Comment by axic** [comment]
It is a though question.  Since we never intended this to happen and rather started out with a very restricted feature set for interfaces (recall #3420 and many other related issues), it is easy to see that this is an oversight and a bug.

However given this likely existed for 2 years and that some forms worked, makes me wonder if anyone relies on it?

And if they do, what does that mean? Likely only that they are locked in to the last version if they insist on this feature.  If we want to be careful, we could ask others from the ecosystem (like the tooling channel).  If we are more confident that this is something nobody ever should have used, then we should go ahead in a non-breaking release and create a blog post to spread the word.

I'm leaning towards not making this a breaking change.

File: test/libsolidity/analysis/FunctionCallGraph.cpp (hunk 1 above)
Comment (cameel) [reply test/libsolidity/analysis/FunctionCallGraph.cpp:1117]: Yeah, though when working on the `FunctionCallGraph` I just tried everything that compiled so it's not like it was by design.

**Comment by chriseth** [comment]
I'm also fine with calling it a bugfix.

**Comment by chriseth** [comment]
The documentation does not mention modifiers, but it says that functions cannot be implemented, so I hope nobody thought that modifiers can be implemented... https://docs.soliditylang.org/en/v0.8.6/contracts.html#interfaces

Can you also update the documentation? It also seems that the bullet list somehow does not render properly.

**Comment by axic** [comment]
> It also seems that the bullet list somehow does not render properly.

I mentioned this on matrix two weeks ago, none of the bullet points render anywhere.

**Comment by cameel** [comment]
Rebased to resolve the changelog conflict and update docs.

> Can you also update the documentation? It also seems that the bullet list somehow does not render properly.

I updated the page to mention modifiers.

The bullet list issue seems to be a general issue with the docs now so I'm going to work on it in a separate PR. No need to hold back this one because of it.

File: docs/contracts/interfaces.rst [hunk 2]
```diff
@@ -6,12 +6,14 @@
 Interfaces
 **********
 
-Interfaces are similar to abstract contracts, but they cannot have any functions implemented. There are further restrictions:
+Interfaces are similar to abstract contracts, but they cannot have any functions or modifiers implemented.
```
Comment (chriseth) [inline docs/contracts/interfaces.rst]: If you say "modifiers implemented" here, it looks like it is OK to declare them.

Comment (cameel) [reply docs/contracts/interfaces.rst]: There's `They cannot declare modifiers` below. I wanted to emphasize that they cannot be defined either.

Comment (cameel) [reply docs/contracts/interfaces.rst]: I reworded it. How about now?

**Comment by chriseth** [comment]
Now it's inconsistent with e.g. constructors, because there you only write "declare".

**Comment by cameel** [comment]
Changed to just "declare".

**Review by Marenz — DISMISSED** [review]

**Review by Marenz — CHANGES_REQUESTED** [review]
LGTM, only needs a rebase and an update to the changelog to add them under the right version.

**Comment by Marenz** [comment]
Rebased & fixed changelog entry position.

**Review by Marenz — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**